polar-core/src/vm.rs

```diff
@@ -2318,44 +2309,16 @@ impl PolarVirtualMachine {
                 }
             }
 
-            // Unify predicates like unifying heads
-            (Value::Call(left), Value::Call(right)) => {
-                // Handled in the parser.
-                assert!(left.kwargs.is_none());
-                assert!(right.kwargs.is_none());
-                if left.name == right.name && left.args.len() == right.args.len() {
-                    self.append_goals(left.args.iter().zip(right.args.iter()).map(
-                        |(left, right)| Goal::Unify {
-                            left: left.clone(),
-                            right: right.clone(),
-                        },
-                    ))?;
-                } else {
-                    self.push_goal(Goal::Backtrack)?
-                }
-            }
-
-            // External instances can unify if they are the same instance, i.e., have the same
-            // instance ID. This is necessary for the case where an instance appears multiple times
-            // in the same rule head. For example, `f(foo, foo) if ...` or `isa(x, y, x: y) if ...`
-            // or `max(x, y, x) if x > y;`.
-            (
-                Value::ExternalInstance(ExternalInstance {
-                    instance_id: left_instance,
-                    ..
-                }),
-                Value::ExternalInstance(ExternalInstance {
-                    instance_id: right_instance,
-                    ..
-                }),
-            ) => {
-                // If IDs match, they're the same _instance_ (not just the same _value_), so unify.
-                if left_instance != right_instance {
-                    self.push_goal(Goal::UnifyExternal {
-                        left_instance_id: *left_instance,
-                        right_instance_id: *right_instance,
-                    })?;
-                }
+            // If either operand is an external instance, let the host
+            // compare them for equality. This handles unification between
+            // "equivalent" host and native types transparently.
+            (Value::ExternalInstance(_), _) | (_, Value::ExternalInstance(_)) => {
```

Comment:
We should keep the previous case that compared both instances by ID to avoid an external event if the operators are both external instances.

Reply:
👍 it's back in